utils/src/priority_set.rs

```diff
@@ -0,0 +1,160 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::hash::Hash;
+
+/// An entry in the `PrioritySet`.
+#[derive(Eq, PartialEq)]
+struct Entry<I: Ord + Hash + Clone, P: Ord + Copy> {
+    item: I,
+    priority: P,
+}
+
+impl<I: Ord + Hash + Clone, P: Ord + Copy> Ord for Entry<I, P> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.priority.cmp(&other.priority) {
+            Ordering::Equal => self.item.cmp(&other.item),
+            other => other,
+        }
+    }
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Copy> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A set that offers fast, priority-ordered iteration over
+/// its elements.
+pub struct PrioritySet<I: Ord + Hash + Clone, P: Ord + Copy> {
+    entries: BTreeSet<Entry<I, P>>,
```

Comment:
While it may look tempting to replace these two inner structures with an "indexed heap", the only object we wouldn't be copying would be `P` (which is expected to be very cheap to copy). I felt it wasn't worth the complexity.